### PR TITLE
Remove experimental flag from visibility triggers and add documentation.

### PR DESCRIPTION
**Pull request by avimehta**
/cc @rudygalfi 

#1297

### CODE REVIEW

**Comment by dvoytenko** [comment]
@avimehta I see experiment itself is removed, but I don't see where it's cleaned.

Assigning to @rudygalfi for docs review.

File: extensions/amp-analytics/amp-analytics.md [hunk 1]
```diff
@@ -319,13 +319,25 @@ The `triggers` attribute describes when an analytics request should be sent. It
   - `timerSpec` (required when `on` is set to `timer`) This configuration is used on conjunction with the `timer` trigger. Please see below for details.
 
 #### Page visible trigger (`"on": "visible"`)
-Use this configuration to fire a request when the page becomes visible. No further configuration is required.
+Use this configuration to fire a request when the page becomes visible. The firing of this trigger can be configured using `visibilitySpec`. Configuration properties supported in `visibilitySpec` are:
+  - `selector` This property can be used to specify the element to which all the `visibilitySpec` conditions apply.
+  - `continuousTimeMin` and `continuousTimeMax` These conditions keep track of the amount of time (any part of) an element has been in the viewport and fire when the time is between the minimum and maximum configured time. The time is expressed in milliseconds.
```
Comment (rudygalfi) [inline extensions/amp-analytics/amp-analytics.md]: These properties indicate that a request should be fired when (any part of) and element has been within the viewport between the minimum and maximum specified times. The times are expressed in milliseconds.

File: extensions/amp-analytics/amp-analytics.md [hunk 2]
```diff
@@ -319,13 +319,25 @@ The `triggers` attribute describes when an analytics request should be sent. It
...
+  - `totalTimeMin` and `totalTimeMax` These conditions keep track of total time for which any part of an element is on screen and fire when the time is between minimum and maximum. The time is expressed in milliseconds.
```
Comment (rudygalfi) [inline extensions/amp-analytics/amp-analytics.md]: These properties indicate that a request should be fired when (any part of) and element has been within the viewport for a total amount of time that is between the minimum and maximum specified times. The times are expressed in milliseconds.

File: extensions/amp-analytics/amp-analytics.md (hunk 1 above)
Comment (rudygalfi) [inline extensions/amp-analytics/amp-analytics.md]: Perhaps even clearer:

These properties indicate that a request should be fired when (any part of) and element has been within the viewport for a continuous amount of time that is between the minimum and maximum specified times. The times are expressed in milliseconds.

File: extensions/amp-analytics/amp-analytics.md [hunk 3]
```diff
@@ -319,13 +319,25 @@ The `triggers` attribute describes when an analytics request should be sent. It
...
+  - `visiblePercentageMin` and `visiblePercentageMax` These conditions keep track of the amount of element visible in vieweport at any given time. They get triggered when the percentage of element visible is between min and max (if specified). Valid values are 0-100 (lower bound is not inclusive while the upper bound is).
```
Comment (rudygalfi) [inline extensions/amp-analytics/amp-analytics.md]: These properties indicate that a request should be fired when the proportion of an element that is visible within the viewport is between the minimum and maximum specified percentages. Percentage values between 0 and 100 are valid. Note that the lower bound (`visiblePercentageMin`) is inclusive while the upper bound (`visiblePercentageMax`) is not.

File: extensions/amp-analytics/amp-analytics.md [hunk 4]
```diff
@@ -319,13 +319,25 @@ The `triggers` attribute describes when an analytics request should be sent. It
...
+
+In addition to the conditions above, `visibilitySpec` also enables certain variables which are documented [here](./analytics-vars.md#visibility-variables).
```
Comment (rudygalfi) [inline extensions/amp-analytics/amp-analytics.md]: Assuming the statement I write here is true, should have something like: "If multiple properties are specified, they must all be true in order for a request to fire."

File: extensions/amp-analytics/analytics-vars.md [hunk 5]
```diff
@@ -42,6 +42,16 @@ When the same `var` is defined in multiple locations, the value is picked in the
 
 The remainder of this doc lists and describes the variables supported in `amp-analytics`.
 
+| Table of Content  |
```
Comment (rudygalfi) [inline extensions/amp-analytics/analytics-vars.md]: Table of contents (plural)

File: extensions/amp-analytics/analytics-vars.md [hunk 6]
```diff
@@ -361,3 +371,75 @@ Provides the number of seconds that have elapsed since 1970. (Epoch time)
 
 Example value: `1452710304312`
 
+## Visibility Variables
+
+### backgrounded
+
+A binary variable with possible values of 1 and 0 to indicate that the page/tab was sent to background at any point before the hit was sent. 1 indicates that the page was backgrounded while 0 indicates that the page has always been in the forground.
```
Comment (rudygalfi) [inline extensions/amp-analytics/analytics-vars.md]: spelling: "foreground"

File: extensions/amp-analytics/analytics-vars.md [hunk 7]
```diff
@@ -361,3 +371,75 @@ Provides the number of seconds that have elapsed since 1970. (Epoch time)
...
+
+### backgroundedAtStart
+
+A binary variable with possible values of 1 and 0 to indicate that the page/tab backgrounded at the time when the page was loaded. 1 indicates that the page was loaded in the background while 0 indicates otherwise.
```
Comment (rudygalfi) [inline extensions/amp-analytics/analytics-vars.md]: "page/tab was backgrounded"

File: extensions/amp-analytics/analytics-vars.md [hunk 8]
```diff
@@ -361,3 +371,75 @@ Provides the number of seconds that have elapsed since 1970. (Epoch time)
 
 Example value: `1452710304312`
 
+## Visibility Variables
+
+### backgrounded
```
Comment (rudygalfi) [inline extensions/amp-analytics/analytics-vars.md]: Are the below two variables related to prerendering or separate. It might be good to note either way. For example, when could a page be backgroundedAtStart?

File: extensions/amp-analytics/analytics-vars.md [hunk 9]
```diff
@@ -361,3 +371,75 @@ Provides the number of seconds that have elapsed since 1970. (Epoch time)
...
+
+A binary variable with possible values of 1 and 0 to indicate that the page/tab was sent to background at any point before the hit was sent. 1 indicates that the page was backgrounded while 0 indicates that the page has always been in the forground.
+
+### backgroundedAtStart
+
+A binary variable with possible values of 1 and 0 to indicate that the page/tab backgrounded at the time when the page was loaded. 1 indicates that the page was loaded in the background while 0 indicates otherwise.
+
+### continuousVisibleTime
```
Comment (rudygalfi) [inline extensions/amp-analytics/analytics-vars.md]: Wondering if it makes sense to call this maximumContinuousVisibleTime because it's the maximum. Theoretically you can also provide the minimum (non-zero) value, too.

File: extensions/amp-analytics/analytics-vars.md [hunk 10]
```diff
@@ -361,3 +371,75 @@ Provides the number of seconds that have elapsed since 1970. (Epoch time)
...
+
+Provides the maximum amount of continuous time an element has met the `visibilitySpec` conditions at the time this ping is sent. Note that a ping with a continuousTimeMin=1000 and totalTimeMin=5000 that is visible for 1000ms, then not visible, then visible
+for 2000ms, then not, then visible for 1000ms, then not, then visible for 1020ms
+will report 2000 for this number as that is the max continuous visible time,
+even if it is not the current continuous visible time (1020 in this example).
```
Comment (rudygalfi) [inline extensions/amp-analytics/analytics-vars.md]: Would currentContinuousVisible time be useful to provide? Seems like it might be.

File: extensions/amp-analytics/analytics-vars.md [hunk 11]
```diff
@@ -361,3 +371,75 @@ Provides the number of seconds that have elapsed since 1970. (Epoch time)
...
+
+Provides the X coordinate of the left edge of the element specified by `visibilitySpec`.
+
+### elementY
+
+Provides the Y coordinate of the top edge of the element specified by `visibilitySpec`.
+
+### firstSeenTime
+
+Provides the time when at least 1px of the element is on the screen for the first time since javascript load.
```
Comment (rudygalfi) [inline extensions/amp-analytics/analytics-vars.md]: Can this be formalized some more. For example, would it be fair to say "since the page became visible"? Or is this really gated on JS load. Does it specifically mean amp-analytics JS?

File: extensions/amp-analytics/analytics-vars.md [hunk 12]
```diff
@@ -361,3 +371,75 @@ Provides the number of seconds that have elapsed since 1970. (Epoch time)
...
+Provides the time when the element met visibility conditions for the last time since
+javascript load.
+
+### loadTimeVisibility
+
+Provides the percentage of element visible in the viewport at load time. This variable assumes that the page is scrolled to top.
+
+### maxVisiblePercentage
+
+Provides the maximum visible percentage over the time that `visibilitySpec` conditions were met. For example, a ping where the element was 100%, then off the page, then 100% will report this value as 100. A ping with visiblePercentageMax=50 undergoing the same transitions would report somewhere between 0 and 50 since any time where the element was 100% on the page would not be counted.
```
Comment (rudygalfi) [inline extensions/amp-analytics/analytics-vars.md]: "any time when"

File: extensions/amp-analytics/analytics-vars.md [hunk 13]
```diff
@@ -361,3 +371,75 @@ Provides the number of seconds that have elapsed since 1970. (Epoch time)
...
+
+### minVisiblePercentage
+
+Provides the minimum visible percentage over the time that visibilitySpec conditions were met. For example, a ping where the element was 100%, then off the page, then 100% will report this value as 0. A ping with visiblePercentageMin=50 condition undergoing the same transitions would report somewhere between 50 and 100.
```
Comment (rudygalfi) [inline extensions/amp-analytics/analytics-vars.md]: Make parallel with above definition by concluding: "since any time when the element was 0% on the page would not be counted."

**Comment by rudygalfi** [comment]
LGTM on docs with a few minor comments

File: extensions/amp-analytics/analytics-vars.md (hunk 8 above)
Comment (avimehta) [inline extensions/amp-analytics/analytics-vars.md]: this is not related to prerendering. Nothing in amp-analytics executes before the page is made visible afaik. Will add some context here.

File: extensions/amp-analytics/analytics-vars.md (hunk 10 above)
Comment (avimehta) [inline extensions/amp-analytics/analytics-vars.md]: Probably. Can you open an issue to track it?

File: extensions/amp-analytics/amp-analytics.md (hunk 1 above)
Comment (avimehta) [inline extensions/amp-analytics/amp-analytics.md]: done.

File: extensions/amp-analytics/amp-analytics.md (hunk 2 above)
Comment (avimehta) [inline extensions/amp-analytics/amp-analytics.md]: done.

File: extensions/amp-analytics/amp-analytics.md (hunk 3 above)
Comment (avimehta) [inline extensions/amp-analytics/amp-analytics.md]: done.

File: extensions/amp-analytics/amp-analytics.md (hunk 4 above)
Comment (avimehta) [inline extensions/amp-analytics/amp-analytics.md]: done.

File: extensions/amp-analytics/analytics-vars.md (hunk 5 above)
Comment (avimehta) [inline extensions/amp-analytics/analytics-vars.md]: done.

File: extensions/amp-analytics/analytics-vars.md (hunk 6 above)
Comment (avimehta) [inline extensions/amp-analytics/analytics-vars.md]: done.

File: extensions/amp-analytics/analytics-vars.md (hunk 7 above)
Comment (avimehta) [inline extensions/amp-analytics/analytics-vars.md]: done.

**Comment by avimehta** [comment]
@rudygalfi You LGTM'ed it but pinging you just in case you want to take a second look.

@dvoytenko This is ready for your review.

**Comment by dvoytenko** [comment]
LGTM from my side.

**Comment by rudygalfi** [comment]
@dvoytenko @avimehta Is this ready to be submitted? It looks like Travis failed.

**Comment by avimehta** [comment]
@rudygalfi looking into the test failures. Will update soon.

**Comment by Gueorgi** [comment]
The call https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/0.1/visibility-impl.js#L243 results in undefined for simple buttons (just text, no image etc.) that are valid in AMP. The next line breaks of course on res.getId(); since res is undefined. I think the same happens for img, links etc. About the only thing that seems to work is <amp-img> with real resource (i.e. if it refers to delete image it still fails). Simple buttons, links and img (not amp-img) etc. should work with this but they don't.
